x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/pipelines/nlp/PipelineConfig.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.inference.pipelines.nlp;
+
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.xpack.ml.inference.pipelines.nlp.tokenizers.BertTokenizer;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+
+public class PipelineConfig implements ToXContentObject {
+
+    public static final ParseField VOCAB = new ParseField("vocab");
+    public static final ParseField TASK_TYPE = new ParseField("task_type");
+    public static final ParseField LOWER_CASE = new ParseField("do_lower_case");
+
+    private static final ObjectParser<PipelineConfig.Builder, Void> STRICT_PARSER = createParser(false);
+    private static final ObjectParser<PipelineConfig.Builder, Void> LENIENT_PARSER = createParser(true);
+
+    private static ObjectParser<PipelineConfig.Builder, Void> createParser(boolean ignoreUnknownFields) {
+        ObjectParser<PipelineConfig.Builder, Void> parser = new ObjectParser<>("pipeline_config",
+            ignoreUnknownFields,
+            Builder::new);
+
+        parser.declareStringArray(Builder::setVocabulary, VOCAB);
+        parser.declareString(Builder::setTaskType, TASK_TYPE);
+        parser.declareBoolean(Builder::setDoLowerCase, LOWER_CASE);
+        return parser;
+    }
+
+    public static PipelineConfig fromXContent(XContentParser parser, boolean lenient) {
+        return lenient ? LENIENT_PARSER.apply(parser, null).build() : STRICT_PARSER.apply(parser, null).build();
+    }
+
+    public static String documentId(String model) {
+        return model + "_pipeline_config";
+    }
+
+    private final TaskType taskType;
+    private final List<String> vocabulary;
```

Comment:
I really like the idea of storing this "pipeline config", though I wish we could keep the nomenclature similar between this and "feature processors". I suppose, this pipeline has multiple processors, etc. :).

But, I do think that storing the vocab directly with a pipeline definition is bad news. Don't multiple models use the same vocab?

Reply:
Broadly yes models either have a cased or uncased vocabulary which could be shared but a fine tuned model or one trained with transfer learning may have additional entries in the vocabulary. I wasted a couple of hours tracking down what I thought was a very subtle bug but in the end I found I had simply mixed up the cased and uncased vocabularies. That kind of error is common, we can eliminate it by having a single config file containing the vocab. 

++ on renaming and using familiar terms from DFA inference